rfd/0113-moderated-file-transfers.md

```diff
@@ -0,0 +1,176 @@
+4. Similar to how we send a `"window-change"` event, we would send a new event 
+	1. `_, err := s.ch.SendRequest("file-transfer-request", false, Marshal(&req))`
+5. Handle the request on the sshserver:
+	1. send an audit event similar to the broadcast above
+	2. store the request in the session
+	3. Notify all members of the party (besides originator) of the request via the session connection. 
+
+The reason we notify _all_ party members (originator excluded) is that a session might not necessarily require a moderator but just a second-party such as a peer. This would allow those types of sessions to still be audited and go through the file request approval process.
+
+After the steps above we are in the `PENDING` state. This isn't a real "state" that exists in any struct, just the part of the flow that is waiting for responses. This is not a blocking state and essentially just creates the request, and updates the relevant UI (more on the UI below). 
+
+#### The Approval process
+
+Similar to sending the `FileTransferRequest` event, another event will be listened to, the `FileTranferRequestResponse` (working name). We could separate this new event into two with Approve/Deny but a simple switch case on a field should suffice. 
+
+OnDeny:
+	1. Remove request from the session struct. No need to keep it around in a "Denied" state, just delete.
+	2. Emit audit event "XUser has denied ~/myfile.txt download request from michael".
+	3. Broadcast message same as audit event above so the rest of the session can see
+	4. Notify all members of the denial so the UI can be removed.
+
+OnApprove:
+	1. Add approver to the `approvers` slice
+	2. Broadcast/audit event
+	3. We can then use a policy checker to see if the approvers fulfill any moderation policy on the original requester. We can treat this check the same as the `checkIfStart` conditional for opening a session. If this comes back true, we notify the original requester with an event containing the ID of the `FileTransferRequest`
+
+Once the client receives this final "approved" message, we can automatically send a "normal" SCP request (over HTTP) with two new optional params, `sessionID` and `requestID` (similar to the new optional `webauthn` param in this same request). The benefits of using the normal SCP request is that we can conditionally choose to skip this entire approval process flow for non-moderated sessions. If the session is not moderated, just send the scp request as usual. If it is, do the song and dance perscribed above.
+
+#### Updated file transfer api handler
+
+Not much will change here. The only difference is adding the `requestID` and `sessionID` as env vars in the `serverContext`. This will allow us to extend our `SessionRegistry.OpenExecSession` method with a new check like `isApprovedRequest`
+
+```go
+func (s *SessionRegistry) isApprovedRequest(ctx context.Context, scx *ServerContext) (bool error) {
```

Comment:
`request` is a vague term and makes me think of access requests here. Maybe `isApprovedFileTransfer` or something along those lines would be a better name.

Reply:
`isApprovedFileTransfer` will work great. If we want, we can even abstract it early (and in other places in the code) to `isApprovedCommand`? its only being implemented with file transfers right now, this process of requesting an command, approving a command, and running that command seems like it can be related to anything we want in the future. Thoughts?